dozer/cogs/filter.py

```diff
@@ -0,0 +1,266 @@
+	@filter.group(invoke_without_command=True)
+	async def whitelist(self, ctx):
+		"""List all whitelisted roles for this server"""
+		with db.Session() as session:
+			results = session.query(WordFilterRoleWhitelist).filter(WordFilterRoleWhitelist.guild_id == ctx.guild.id).all()
+			roles_text = ""
+			for db_role in results:
+				role = discord.utils.get(ctx.guild.roles, id=db_role.role_id)
+				roles_text += "{}\n".format(role.name)
+			embed = discord.Embed()
+			embed.title = "Whitlisted roles for {}".format(ctx.guild.name)
+			embed.description = "Anybody with any of the roles below will not have their messages filtered."
+			embed.add_field(name="Roles",value=roles_text or "No roles")
+			await ctx.send(embed=embed)
+	whitelist.example_usage = "{prefix}filter whitelist"
+
+	@guild_only()
+	@has_permissions(manage_roles=True)
+	@whitelist.command(name="add")
+	async def whitelist_add(self, ctx, *, role: discord.Role):
+		"""Add a role to the whitelist"""
+		with db.Session() as session:
+			result = session.query(WordFilterRoleWhitelist).filter(WordFilterRoleWhitelist.role_id == role.id).one_or_none()
+			if result is not None:
+				await ctx.send("That role is already whitelisted.")
+			whitelist_entry = WordFilterRoleWhitelist(guild_id=ctx.guild.id, role_id=role.id)
+			session.add(whitelist_entry)
+			await ctx.send("Whitelisted `{}` for this guild.".format(role.name))
+	whitelist_add.example_usage = "{prefix}filter whitelist add Moderators"
+
+	@guild_only()
+	@has_permissions(manage_roles=True)
+	@whitelist.command(name="remove")
+	async def whitelist_remove(self, ctx, *, role:discord.Role):
+		"""Remove a role from the whitelist"""
+		with db.Session() as session:
+			result = session.query(WordFilterRoleWhitelist).filter(WordFilterRoleWhitelist.role_id == role.id).one_or_none()
+			if result is None:
+				await ctx.send("That role is not whitelisted.")
+			session.delete(result)
+			await ctx.send("The role `{}` is no longer whitelisted.".format(role.name))
+	whitelist_remove.example_usage = "{prefix}filter whitelist remove Admins"
+
+
+def setup(bot):
+	bot.add_cog(Filter(bot))
+
+
+"""Database Tables"""
+
+
+class WordFilter(db.DatabaseObject):
+	__tablename__ = "word_filters"
+	id = db.Column(db.Integer, primary_key=True)
+	enabled = db.Column(db.Boolean, default=True)
+	guild_id = db.Column(db.Integer)
+	friendly_name = db.Column(db.String, nullable=True)
+	pattern = db.Column(db.String)
+	infractions = db.relationship("WordFilterInfraction", back_populates="filter")
+
+
+class WordFilterSetting(db.DatabaseObject):
+	__tablename__ = "word_filter_settings"
+	id = db.Column(db.Integer, primary_key=True)
+	setting_type = db.Column(db.String)
+	guild_id = db.Column(db.Integer)
+	value = db.Column(db.String)
+
+
+class WordFilterRoleWhitelist(db.DatabaseObject):
+	__tablename__ = "word_filter_role_whitelist"
+	id = db.Column(db.Integer, primary_key=True)
+	guild_id = db.Column(db.Integer)
+	role_id = db.Column(db.Integer)
```

Comment:
Role IDs are unique and this is essentially a representation of a role in the database, so using the role ID as the primary key looks reasonable to me.